File: src/routes/apps.ts
import type { FastifyInstance } from "fastify";

import { decode } from "jsonwebtoken";
import { parseAuthorizationHeader } from "../util/auth";
import { db } from "../util/db";
import { redis } from "../util/redis";

interface GetAppChannels {
  Params: { appId: string };
  Querystring: { filterByPrefix?: string; cursor?: string };
}

export async function routes(server: FastifyInstance) {
  server.get<GetAppChannels>("/:appId/channels", async (req, reply) => {
    const { appId } = req.params;
    let token;
    try {
      token = parseAuthorizationHeader(req.headers.authorization);
    } catch (error) {
      return reply.status(401).send({
        errors: [(error as Error).message],
      });
    }

    let jwt;
    try {
      jwt = decode(token, { complete: true });
    } catch (error) {
      return reply.status(401).send({
        errors: ["Invalid authorization token"],
      });
    }

    if (!jwt) {
      return reply.status(401).send({
        errors: ["Invalid authorization token"],
      });
    }

    const { kid: apiKeyId } = jwt.header;
    if (!apiKeyId) {
      return reply.status(401).send({
        errors: ["Invalid authorization token"],
      });
    }

    const apiKey = await db.apiKey.findFirst({
      where: {
        AND: [{ id: { equals: apiKeyId } }, { appId: { equals: appId } }],
      },
      select: {
        id: true,
        secret: true,
        capabilities: true,
        app: {
          select: {
            id: true,
          },
        },
      },
    });

    if (!apiKey) {
      return reply.status(401).send({
        errors: ["Invalid authorization token"],
      });
    }

    const filterByPrefix = req.query.filterByPrefix;
    const cursor = Number(req.query.cursor || 0);
    let match: string;
    if (filterByPrefix) {
      match = `subscribers:${apiKey.app.id}:${filterByPrefix}*`;
    } else {
      match = `subscribers:${apiKey.app.id}:*`;
    }

    const [updatedCursor, maybeDuplicateChannels] = await redis.scan(
      `${cursor}`,
      "MATCH",
      match,
      "COUNT",
      100
    );

    // There may be duplicate channels due to redis scan so we have to deduplicate.
    const channels = new Set(maybeDuplicateChannels);

    reply.send({
      channels: [...channels].map((channel) => {
        const [, name] = channel.split(`subscribers:${appId}:`);
        return name;
      }),
      cursor: Number(updatedCursor) === 0 ? undefined : updatedCursor,
    });
  });
}

File: pkg/controllers/apps.go
package controllers

import (
	"errors"
	"net/http"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/gmencz/mycelium/pkg/common"
	"github.com/gmencz/mycelium/pkg/models"
	"github.com/golang-jwt/jwt/v4"
)

func (c *Controller) GetApp(ctx *gin.Context) {
	id := ctx.Param("id")
	auth, authErr := common.ParseAuthorizationHeader(ctx.Request.Header.Get("authorization"))
	if authErr != nil {
		ctx.JSON(http.StatusUnauthorized, gin.H{
			"message": authErr.Error(),
		})
		return
	}

	var apiKey models.ApiKey
	_, jwtErr := jwt.Parse(auth, func(token *jwt.Token) (interface{}, error) {
		// Don't forget to validate the alg is what you expect:
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, errors.New("invalid token, unexpected signing method")
		}

		kid := token.Header["kid"]
		if kid == nil {
			return nil, errors.New("invalid token, missing kid header")
		}

		if result := c.Db.First(&apiKey, "id = ?", kid); result.Error != nil {
			return nil, errors.New("invalid token kid header")
		}

		return []byte(apiKey.Secret), nil
	})

	if jwtErr != nil {
		ctx.JSON(http.StatusUnauthorized, gin.H{
			"message": jwtErr.Error(),
		})
		return
	}

	if id != apiKey.AppID {
		ctx.JSON(http.StatusUnauthorized, gin.H{
			"message": "unauthorized to query this app",
		})
		return
	}

	filterByPrefix := ctx.Request.URL.Query().Get("filter_by_prefix")
	cursor := 0

	if ctx.Request.URL.Query().Has("cursor") {
		cursorInt, cursorErr := strconv.Atoi(ctx.Request.URL.Query().Get("cursor"))
		if cursorErr != nil {
			ctx.JSON(http.StatusUnauthorized, gin.H{
				"message": "invalid cursor, must be a number",
			})
			return
		}

		cursor = cursorInt
	}

	var match string
	if filterByPrefix != "" {
		match = "subscribers:" + apiKey.AppID + ":" + filterByPrefix + "*"
	} else {
		match = "subscribers:" + apiKey.AppID + ":*"
	}

	scanCmd := c.Rdb.Scan(ctx, uint64(cursor), match, 100)
	if scanCmd.Err() != nil {
		ctx.JSON(http.StatusInternalServerError, gin.H{
			"message": "internal server error",
		})
		return
	}

	channelsResult, resultCursor, _ := scanCmd.Result()

	uniqueChannels := common.RemoveDuplicateStrings(channelsResult)

	channels := common.Map(uniqueChannels, func(channel string) string {
		channelParts := strings.Split(channel, "subscribers:"+apiKey.AppID+":")
		return channelParts[1]
	})

	if resultCursor != 0 {
		ctx.JSON(http.StatusOK, gin.H{
			"channels": channels,
			"cursor":   resultCursor,
		})

		return
	}

	ctx.JSON(http.StatusOK, gin.H{
		"channels": channels,
	})
}
